src/mlpack/core/tree/cosine_tree/cosine_tree.cpp

```diff
@@ -150,6 +150,96 @@ CosineTree::CosineTree(const arma::mat& dataset,
   ConstructBasis(treeQueue);
 }
 
+//! Copy the given tree.
+CosineTree::CosineTree(const CosineTree& other) :
+  dataset((other.parent == NULL) ? other.dataset : NULL),
+  delta(other.delta),
+  parent(other.Parent()),
+  left(other.Left()),
+  right(other.Right()),
+  indices(other.indices),
+  l2NormsSquared(other.l2NormsSquared),
+  centroid(other.centroid),
+  basisVector(other.basisVector),
+  splitPointIndex(other.SplitPointIndex()),
+  numColumns(other.NumColumns()),
+  l2Error(other.L2Error()),
+  frobNormSquared(other.FrobNormSquared())
+{
+  //copy children
+}
+
+//! Copy Assignment
+CosineTree&
+CosineTree::
+operator=(const CosineTree& other)
+{
+  // Return if it's the same tree.
+  if (this == &other)
+    return *this;
+
+  //dataset = (other.parent == NULL) ? other.dataset : NULL;
```

Comment:
Dataset is a reference over here not a pointer. :)

Reply:
I'm not sure I follow. I saw that the datatype is `const arma::mat` instead of `MatType*` like it is for the other trees. I read up on C++ references but I can't figure out why I can't assign the dataset reference. 
How do I go about assigning a reference?

Reply:
https://www.geeksforgeeks.org/when-do-we-use-initializer-list-in-c/
This will answer your doubt. 2nd point.

Reply:
Here's the correct code for making a perfect copy (deep copy)
use the initializer list to remove the error. 
dataset (other.getdataset())

Then in the constructor body 
```
arma::mat a = other.getDataset();

this->dataset = a;
```

Reply:
It's still not working. I've changed `dataset((other.parent == NULL) ? other.dataset : NULL)` to `dataset((other.parent == NULL) ? other.GetDataset() : NULL)` in the Copy Constructor's initializer list.

> Then in the constructor body

Did you mean to say in the body of Copy Assignment? Nevertheless, I tried adding the two lines you've specified in both the body of the Constructor and the Assignment but the error didn't go away.

Reply:
> It's still not working. I've changed `dataset((other.parent == NULL) ? other.dataset : NULL)` to `dataset((other.parent == NULL) ? other.GetDataset() : NULL)` in the Copy Constructor's initializer list.

What's the error the initializer list is throwing?

Reply:
The initializer list doesn't show any error, with or without GetDataset(). It's the assignment in Copy and Move assignment which shows the error.

The error is `no operator "=" matches these operands -- operand types are: const arma::mat = const arma::mat`

Reply:
Okay fine. That error is because of using a `const` object reference as it will need a `const` function. So I would say it will be good if we make `dataset` member as `non-const reference` inside `cosine_tree.hpp`. 
However, the author would have kept it `const` because of some reason. So let us know if it creates an error again after making the said change.